File: Week5/BoilerCodes/ArrayVariation/stack.py
class Stack:
    def __init__(self):
        self.MAX = 50
        self.items = [0] * self.MAX
        self.top = -1

    def is_empty(self):
        return self.top == -1

    def is_full(self):
        return self.top == self.MAX - 1

    def push(self, item):
        if self.is_full():
            print(f"Stack Overflow! Cannot push {item}")
            return False
        self.top += 1
        self.items[self.top] = item
        return True

    def pop(self):
        if self.is_empty():
            print("Stack Underflow! Cannot pop from empty stack")
            return -1
        item = self.items[self.top]
        self.top -= 1
        return item

    def peek(self):
        if self.is_empty():
            print("Stack is empty! Cannot peek")
            return -1
        return self.items[self.top]

File: Week5/BoilerCodes/ArrayVariation/stack.go
package main

import "fmt"

const MAX = 50

type Stack struct {
	items [MAX]int
	top   int
}

func NewStack() *Stack {
	return &Stack{
		top: -1,
	}
}

func (s *Stack) IsEmpty() bool {
	return s.top == -1
}

func (s *Stack) IsFull() bool {
	return s.top == MAX-1
}

func (s *Stack) Push(item int) bool {
	if s.IsFull() {
		fmt.Printf("Stack Overflow! Cannot push %d\n", item)
		return false
	}
	s.top++
	s.items[s.top] = item
	return true
}

func (s *Stack) Pop() int {
	if s.IsEmpty() {
		fmt.Println("Stack Underflow! Cannot pop from empty stack")
		return -1
	}
	item := s.items[s.top]
	s.top--
	return item
}

func (s *Stack) Peek() int {
	if s.IsEmpty() {
		fmt.Println("Stack is empty! Cannot peek")
		return -1
	}
	return s.items[s.top]
}
